### PR TITLE
types,expression: make 'invalid time format' error log more friendly

### PR DESCRIPTION
**Pull request by tiancaiamao**
From 'invalid time format', we have no idea why the data is wrong.
So I change 'invalid time format' to 'invalid time format: %s'

@XuHuaiyu @coocood @zz-jason

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -1368,13 +1368,13 @@ func checkDatetimeType(t MysqlTime, allowZeroInDate bool) error {
 
 	hour, minute, second := t.Hour(), t.Minute(), t.Second()
 	if hour < 0 || hour >= 24 {
-		return errors.Trace(ErrInvalidTimeFormat)
+		return errors.Trace(ErrInvalidTimeFormat.GenByArgs(hour))
 	}
 	if minute < 0 || minute >= 60 {
-		return errors.Trace(ErrInvalidTimeFormat)
+		return errors.Trace(ErrInvalidTimeFormat.GenByArgs(minute))
```
Comment (zimulala) [inline types/time.go]: Do we need to add the minute field in the log like `fmt.Sprintf("minute: %v", minute)` or use `t` as the arg ?

Comment (tiancaiamao) [reply types/time.go]: The error stack log would have enough information the trace to this line, so the minute value is enough

**Comment by zimulala** [comment]
Do we need to update mysql_test?

**Comment by tiancaiamao** [comment]
Let me have a try. 
/run-all-tests

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: types/convert.go [hunk 2]
```diff
@@ -217,7 +217,7 @@ func NumberToDuration(number int64, fsp int) (Duration, error) {
 	}
 
 	if number/10000 > TimeMaxHour || number%100 >= 60 || (number/100)%100 >= 60 {
-		return ZeroDuration, ErrInvalidTimeFormat
+		return ZeroDuration, errors.Trace(ErrInvalidTimeFormat.GenByArgs(number))
```
Comment (zz-jason) [inline types/convert.go]: no need to `Trace`, it's already the root cause.

File: types/mytime.go [hunk 3]
```diff
@@ -112,7 +112,7 @@ func (t MysqlTime) GoTime(loc *gotime.Location) (gotime.Time, error) {
 	if year != t.Year() || int(month) != t.Month() || day != t.Day() ||
 		hour != t.Hour() || minute != t.Minute() || second != t.Second() ||
 		microsec != t.Microsecond() {
-		return tm, errors.Trace(ErrInvalidTimeFormat)
+		return tm, errors.Trace(ErrInvalidTimeFormat.GenByArgs(t))
```
Comment (zz-jason) [inline types/mytime.go]: ditto, and so as to the others.

File: types/convert.go (hunk 2 above)
Comment (tiancaiamao) [reply types/convert.go]: I'm afraid this is necessary, without it we'll lost the last stack

```
package main

import (
	"fmt"

	"github.com/juju/errors"
)

func main() {
	e := h()
	fmt.Println(errors.ErrorStack(e))
}

type MyE struct{}

func (_ MyE) Error() string {
	return "hello"
}

func h() error {
	return errors.Trace(g())
}

func g() error {
	return errors.Trace(f())
}

func f() error {
	if false {
		return MyE{}  
	}
	return MyE{} 
}
```

The result is:

```
➜  test go run test.go
hello
/home/genius/project/test/test.go:25: 
/home/genius/project/test/test.go:21: 
```

f() may return two errors, we don't know which one without line information.
Change f() to 

```
func f() error {
	if false {
		return errors.Trace(MyE{})
	}
	return errors.Trace(MyE{})
}
```

```
➜  test go run test.go
hello
/home/genius/project/test/test.go:32: 
/home/genius/project/test/test.go:25: 
/home/genius/project/test/test.go:21:
```

We know exactly it's line 32, not line 30...

Comment (zz-jason) [reply types/convert.go]: ok

**Pull request merged**